src/FshCompletionProvider.ts

```diff
@@ -0,0 +1,116 @@
+import {
+  CompletionItemProvider,
+  TextDocument,
+  Position,
+  ProviderResult,
+  CompletionItem,
+  CompletionList,
+  Range
+} from 'vscode';
+import { EntityType, FshDefinitionProvider } from './FshDefinitionProvider';
+
+export class FshCompletionProvider implements CompletionItemProvider {
+  constructor(private definitionProvider: FshDefinitionProvider) {}
+
+  public getAllowedTypesAndExtraNames(
+    document: TextDocument,
+    position: Position
+  ): {
+    allowedTypes: EntityType[];
+    extraNames: CompletionItem[];
+  } {
+    // search backwards from our current position to get the right kinds of names
+    // check the nearest non-empty line. don't just keep going forever
+    // if we see "InstanceOf" we're making an instance, which means we want Profile, Resource, or Extension
+    // if we see "obeys" we're setting an obeysRule, which means we want Invariant
+    // if we see "Parent" keep looking, lots of things have parents
+    // A Profile's parent can be a Profile, Logical,
+    // An Extension's parent must also be an Extension
+    // A Logical's parent must be a Logical or a Resource, or "Base" or "Element"
+    // A Resource's parent must be "Resource" or "DomainResource"
+
+    const currentLine = document.getText(
+      new Range(position.line, 0, position.line, position.character)
+    );
+    let allowedTypes: EntityType[] = [];
+    const extraNames: CompletionItem[] = [];
+    if (currentLine.startsWith('InstanceOf:')) {
+      allowedTypes = ['Profile', 'Resource', 'Extension'];
+    } else if (currentLine.match(/^[ \t]*\* (\S+ )?obeys/)) {
+      allowedTypes = ['Invariant'];
+    } else if (currentLine.startsWith('Parent:')) {
+      // find the most recent non-empty line
+      let previousLineNumber: number;
+      for (previousLineNumber = position.line - 1; previousLineNumber >= 0; previousLineNumber--) {
+        const previousLine = document
+          .getText(new Range(previousLineNumber, 0, previousLineNumber + 1, 0))
+          .trim();
+        if (previousLine.length > 0) {
+          if (previousLine.match(/Profile:/)) {
+            allowedTypes = ['Profile', 'Logical'];
```

Comment:
I'm not sure if it's too much of an edge case, but it looks like SUSHI allows a Profile to have an Extension as the parent and a Resource as the parent. Is that something we want to support in the autocomplete as well?